File: clojure/src/trieshake/scanner.clj
(ns trieshake.scanner
  "Scanner module — walk a directory tree, collect files matching an extension."
  (:require [clojure.java.io :as io]
            [clojure.string :as str]))

(defn scan-files
  "Walk base-dir and collect files matching the extension.
   Returns a sorted seq of {:abs-path abs :rel-path rel} maps."
  [base-dir extension]
  (let [base (io/file base-dir)
        base-path (.toPath base)
        all-files (file-seq base)]
    (->> all-files
         (filter #(.isFile %))
         ;; Skip hidden files (any path component starting with .)
         (remove (fn [f]
                   (let [rel (.relativize base-path (.toPath f))
                         parts (iterator-seq (.iterator rel))]
                     (some #(str/starts-with? (str %) ".") parts))))
         ;; Filter by extension (case-insensitive, suffix-based)
         (filter (fn [f]
                   (if extension
                     (str/ends-with? (str/lower-case (.getName f))
                                     (str/lower-case extension))
                     true)))
         ;; Build result maps
         (map (fn [f]
                {:abs-path (.getAbsolutePath f)
                 :rel-path (str (.relativize base-path (.toPath f)))}))
         (sort-by :rel-path)
         vec)))

File: trieshake/scanner.py
"""Scanner module — walk a directory tree, collect files matching an extension."""

from pathlib import Path


def scan_files(
    base_dir: Path, extension: str | None = None
) -> list[tuple[Path, Path]]:
    """Walk base_dir and collect files matching the extension.

    Args:
        base_dir: The root directory to scan.
        extension: File extension to match (case-insensitive, suffix-based).
                   If None, match all files.

    Returns:
        List of (absolute_path, relative_path) tuples, sorted by relative path.
    """
    results = []
    for abs_path in sorted(base_dir.rglob("*")):
        if not abs_path.is_file():
            continue
        # Skip hidden files
        if any(part.startswith(".") for part in abs_path.relative_to(base_dir).parts):
            continue
        if extension is not None:
            if not abs_path.name.lower().endswith(extension.lower()):
                continue
        rel_path = abs_path.relative_to(base_dir)
        results.append((abs_path, rel_path))
    return results
